subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayApplicationRunner.java

```diff
@@ -16,35 +16,102 @@
 import org.gradle.process.internal.worker.WorkerProcessFactory;
 
 import java.io.File;
+import java.util.Set;
 
 public class PlayApplicationRunner {
     private final WorkerProcessFactory workerFactory;
     private final VersionedPlayRunAdapter adapter;
+    private ClasspathSnapshotter snapshotter;
 
-    public PlayApplicationRunner(WorkerProcessFactory workerFactory, VersionedPlayRunAdapter adapter) {
+    public PlayApplicationRunner(WorkerProcessFactory workerFactory, VersionedPlayRunAdapter adapter, ClasspathSnapshotter snapshotter) {
         this.workerFactory = workerFactory;
         this.adapter = adapter;
+        this.snapshotter = snapshotter;
     }
 
     public PlayApplication start(PlayRunSpec spec, Deployment deployment) {
         WorkerProcess process = createWorkerProcess(spec.getProjectPath(), workerFactory, spec, adapter);
         process.start();
 
         PlayRunWorkerServerProtocol workerServer = process.getConnection().addOutgoing(PlayRunWorkerServerProtocol.class);
-        PlayApplication playApplication = new PlayApplication(deployment, workerServer, process);
+        PlayApplication playApplication = new PlayApplication(new PlayClassloaderMonitorDeploymentDecorator(deployment, spec, adapter), workerServer, process);
         process.getConnection().addIncoming(PlayRunWorkerClientProtocol.class, playApplication);
         process.getConnection().connect();
         playApplication.waitForRunning();
         return playApplication;
     }
 
+    private class PlayClassloaderMonitorDeploymentDecorator implements Deployment {
+        private Deployment delegate;
+        private FileCollection applicationClasspath;
+        private HashCode snapshot;
+        private boolean isPlay22;
+
+        private PlayClassloaderMonitorDeploymentDecorator(Deployment delegate, PlayRunSpec runSpec, VersionedPlayRunAdapter adapter) {
+            this.delegate = delegate;
+            this.applicationClasspath = collectApplicationClasspath(runSpec);
+            this.isPlay22 = adapter instanceof PlayRunAdapterV22X;
+        }
+
+        private FileCollection collectApplicationClasspath(PlayRunSpec runSpec) {
+            Set<File> applicationClasspath = Sets.newLinkedHashSet(runSpec.getChangingClasspath());
+            applicationClasspath.add(runSpec.getApplicationJar());
+            return new SimpleFileCollection(applicationClasspath);
+        }
+
+        @Override
+        public Status status() {
+            final Status delegateStatus = delegate.status();
+
+            if (isPlay22) {
+                // PlayRunAdapterV22X doesn't load assets from directory directly
+                return delegateStatus;
+            }
+
+            if (!delegateStatus.hasChanged()) {
+                return delegateStatus;
+            }
+
+            HashCode oldSnapshot = updateSnapshot();
+
+            if (!snapshot.equals(oldSnapshot)) {
+                return delegateStatus;
+            } else {
```

Comment:
For readability I'd extract an `applicationClasspathChanged()` method that hides the implementation details of snapshotting, and change this to:
```
if (!applicationClasspathChanged()) {
    return delegateStatus
}
return new Status() {
   ...
```

Reply:
Good catch, that's much neater.